### PR TITLE
Use gorilla/mux to route HTTP requests

### PR DESCRIPTION
**Pull request by Kegsay**
- Make `/sync` and `/send` use the right API paths.
- Pluck out the `roomID` as an example of how to do it.

### CODE REVIEW

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go [hunk 1]
```diff
@@ -14,7 +15,9 @@ type SendMessage struct {
 // OnIncomingRequest implements util.JSONRequestHandler
 func (s *SendMessage) OnIncomingRequest(req *http.Request) (interface{}, *util.HTTPError) {
 	logger := req.Context().Value(util.CtxValueLogger).(*log.Entry)
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: I realise this isn't part of the PR but would it make sense create a type safe accessors for the logger as per the suggestions in https://golang.org/pkg/context/

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go [hunk 2]
```diff
@@ -14,7 +15,9 @@ type SendMessage struct {
...
-	logger.Info("Doing stuff...")
+	vars := mux.Vars(req)
+	roomID := vars["roomID"]
+	logger.WithField("roomID", roomID).Info("Doing stuff...")
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: You should also grab the eventType from vars here.

Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: The implementation of `mux.Vars` is "interesting". There should probably be a comment saying that this handler has to be called through a gorilla Mux in order for it to have stashed the route vars on the request context.

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go (hunk 1 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: Failing to see where that is mentioned.

File: src/github.com/matrix-org/dendrite/clientapi/routing/routing.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+// Setup registers HTTP handlers with the given ServeMux. It also supplies the given http.Client
+// to clients which need to make outbound HTTP requests.
+func Setup(servMux *http.ServeMux, httpClient *http.Client) {
+	apiMux := mux.NewRouter()
+	r0mux := apiMux.PathPrefix(pathPrefixR0).Subrouter()
+	r0mux.Handle("/sync", make("sync", &readers.Sync{}))
+	r0mux.Handle("/rooms/{roomID}/send/{eventType}", make("send_message", &writers.SendMessage{}))
+
+	servMux.Handle("/metrics", prometheus.Handler())
+	servMux.Handle("/api/", http.StripPrefix("/api", apiMux))
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/clientapi/routing/routing.go]: I like having the paths all in one place. It makes it much easier to see where a request is going to be handled. :+1:

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go (hunk 2 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: It is "interesting" and I'm wondering how best to handle that going forwards. In all practicalities we WILL be using a gorilla mux so I don't feel inclined to doc this every time we pluck something out of vars.

I dislike how separated the path pattern is from where we access things, but I like how we can easily see the endpoints available in routing.go

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go (hunk 1 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]:
```go
     // Packages that define a Context key should provide type-safe accessors
     // for the values stored using that key:
```

Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: Oh okay I'll add it to the util package then.

Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: https://github.com/matrix-org/util/pull/2

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go (hunk 2 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: > You should also grab the eventType from vars here.

Done.

**Comment by Kegsay** [comment]
@NegativeMjark PTAL

**Comment by Kegsay** [comment]
@NegativeMjark PTAL

File: src/github.com/matrix-org/dendrite/clientapi/routing/routing.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+func Setup(servMux *http.ServeMux, httpClient *http.Client) {
+	apiMux := mux.NewRouter()
+	r0mux := apiMux.PathPrefix(pathPrefixR0).Subrouter()
+	r0mux.Handle("/sync", make("sync", wrap(func(req *http.Request) (interface{}, *util.HTTPError) {
+		return readers.Sync(req)
+	})))
+	r0mux.Handle("/rooms/{roomID}/send/{eventType}",
+		make("send_message", wrap(func(req *http.Request) (interface{}, *util.HTTPError) {
+			vars := mux.Vars(req)
+			roomID := vars["roomID"]
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/clientapi/routing/routing.go]: Can we inline the vars["<key>"] here? It doesn't look like we need to store these as local variables here.

**Comment by NegativeMjark** [comment]
Other than maybe inlining the vars[""]. LGTM

File: src/github.com/matrix-org/dendrite/clientapi/routing/routing.go (hunk 4 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/routing/routing.go]: Done.

**Pull request merged**